libyul/optimiser/RedundantAssignEliminator.cpp

```diff
@@ -104,35 +104,25 @@ void RedundantAssignEliminator::operator()(Switch const& _switch)
 
 void RedundantAssignEliminator::operator()(FunctionDefinition const& _functionDefinition)
 {
-	std::set<YulString> outerDeclaredVariables;
-	std::set<YulString> outerReturnVariables;
-	TrackedAssignments outerAssignments;
-	ForLoopInfo forLoopInfo;
-	swap(m_declaredVariables, outerDeclaredVariables);
-	swap(m_returnVariables, outerReturnVariables);
-	swap(m_assignments, outerAssignments);
-	swap(m_forLoopInfo, forLoopInfo);
+	ScopedSaveAndRestore outerDeclaredVariables(m_declaredVariables, {});
+	ScopedSaveAndRestore outerReturnVariables(m_returnVariables, {});
+	ScopedSaveAndRestore outerAssignments(m_assignments, {});
+	ScopedSaveAndRestore forLoopInfo(m_forLoopInfo, {});
 
 	for (auto const& retParam: _functionDefinition.returnVariables)
 		m_returnVariables.insert(retParam.name);
 
 	(*this)(_functionDefinition.body);
 
 	for (auto const& param: _functionDefinition.parameters)
-		finalize(param.name, State::Unused);
+		finalize(param.name, UseState::Unused);
 	for (auto const& retParam: _functionDefinition.returnVariables)
-		finalize(retParam.name, State::Used);
-
-	swap(m_declaredVariables, outerDeclaredVariables);
-	swap(m_returnVariables, outerReturnVariables);
-	swap(m_assignments, outerAssignments);
-	swap(m_forLoopInfo, forLoopInfo);
+		finalize(retParam.name, UseState::Used);
 }
 
 void RedundantAssignEliminator::operator()(ForLoop const& _forLoop)
 {
-	ForLoopInfo outerForLoopInfo;
-	swap(outerForLoopInfo, m_forLoopInfo);
+	ScopedSaveAndRestore outerForLoopInfo(m_forLoopInfo, {});
 	++m_forLoopNestingDepth;
```

Comment:
```suggestion
	ScopedSaveAndRestore loopNestingDepthIncrement(m_forLoopNestingDepth, m_forLoopNestingDepth + 1);
```
not sure that's really nicer, but it's at least more local than having the decrement at the very end. But also fine keeping as is.